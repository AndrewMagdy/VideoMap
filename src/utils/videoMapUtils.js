import Config from "react-native-config";

const apiEndPoint = Config.API_URL;
const apiKey = Config.API_KEY;
const locationRadius = "100km";
const maxResults = 10;

export const getVideosAPI = async (coordinate, pageToken) => {
  const coordinateStr = encodeURIComponent(
    `${coordinate.latitude},${coordinate.longitude}`
  );
  const getRequest = `${apiEndPoint}?part=snippet&location=${coordinateStr}&pageToken=${pageToken}&locationRadius=${locationRadius}&type=video&key=${apiKey}&maxResults=${maxResults}`;
  const response = await fetch(getRequest);
  const responseBody = await response.json();
  if (!response.ok) {
    throw responseBody.error;
  }
  return responseBody;
};
