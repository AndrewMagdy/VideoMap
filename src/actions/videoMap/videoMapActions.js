import { getVideosAPI } from "../../utils/videoMapUtils";
import {
  LOAD_VIDEOS_REQUEST,
  LOAD_VIDEOS_SUCCESS,
  LOAD_VIDEOS_ERROR
} from "../types";

const loadVideosRequest = () => ({
  type: LOAD_VIDEOS_REQUEST
});

const loadVideosSuccess = (videosList, isNextPage) => ({
  type: LOAD_VIDEOS_SUCCESS,
  payload: { videosList, isNextPage }
});

const loadVideosError = error => ({
  type: LOAD_VIDEOS_ERROR,
  payload: { error: error.message }
});

export const loadVideos = (coordinate, pageToken = "") => async dispatch => {
  dispatch(loadVideosRequest());
  try {
    const isNextPage = pageToken === "" ? false : true;
    const loadedVideos = await getVideosAPI(coordinate, pageToken);
    return dispatch(loadVideosSuccess(loadedVideos, isNextPage));
  } catch (error) {
    return dispatch(loadVideosError(error));
  }
};
