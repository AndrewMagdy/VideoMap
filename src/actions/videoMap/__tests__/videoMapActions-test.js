import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { loadVideos } from "../videoMapActions";
import * as videoMapUtils from "../../../utils/videoMapUtils";

describe("VideoMap action creators", () => {
  const mockStore = configureMockStore([thunk])({});

  afterEach(() => {
    mockStore.clearActions();
  });

  const videosAPIResponse = {
    kind: "youtube#searchListResponse",
    etag: '"XpPGQXPnxQJhLgs6enD_n8JR4Qk/P0Q5iJsjiJ2zOzcrxDcb85guSJ8"',
    nextPageToken: "CAIQAA",
    regionCode: "EG",
    pageInfo: {
      totalResults: 1494,
      resultsPerPage: 2
    },
    items: [
      {
        kind: "youtube#searchResult",
        etag: '"XpPGQXPnxQJhLgs6enD_n8JR4Qk/ModJNYQuE9VfOulIBgXRd127_Ew"',
        id: {
          kind: "youtube#video",
          videoId: "ei4r_Bs3cek"
        },
        snippet: {
          publishedAt: "2018-04-26T07:43:02.000Z",
          channelId: "UCVo06dBGK9VBBhq15wJxZHQ",
          title: "#JEFFHUBBARD - #BODYBOARDING #PIPELINE SPECIALIST $$",
          description:
            "Video/Edit: Chris Kincade FINE ART PRINTS www.ChrisKincade.com GET 25% OFF THE ENTIRE SITE! USE CODE “2018” AT CHECKOUT! Instagram: @Chris.",
          thumbnails: {
            default: {
              url: "https://i.ytimg.com/vi/ei4r_Bs3cek/default.jpg",
              width: 120,
              height: 90
            },
            medium: {
              url: "https://i.ytimg.com/vi/ei4r_Bs3cek/mqdefault.jpg",
              width: 320,
              height: 180
            },
            high: {
              url: "https://i.ytimg.com/vi/ei4r_Bs3cek/hqdefault.jpg",
              width: 480,
              height: 360
            }
          },
          channelTitle: "Chris Kincade Media",
          liveBroadcastContent: "none"
        }
      },
      {
        kind: "youtube#searchResult",
        etag: '"XpPGQXPnxQJhLgs6enD_n8JR4Qk/Cizv5T8I6oJtAxERYXtftJ0L3xg"',
        id: {
          kind: "youtube#video",
          videoId: "D6zek91qfoM"
        },
        snippet: {
          publishedAt: "2015-02-02T05:51:33.000Z",
          channelId: "UCVo06dBGK9VBBhq15wJxZHQ",
          title: "#662 Boys take XL #Pipeline - #Bodyboarding #Hawaii",
          description:
            "When Pipe gets too big, it's always the 662 Ride Shop boys that tackle it anyways. Here's some clips from January 22nd of Andre Botha, Trevor Kam and ...",
          thumbnails: {
            default: {
              url: "https://i.ytimg.com/vi/D6zek91qfoM/default.jpg",
              width: 120,
              height: 90
            },
            medium: {
              url: "https://i.ytimg.com/vi/D6zek91qfoM/mqdefault.jpg",
              width: 320,
              height: 180
            },
            high: {
              url: "https://i.ytimg.com/vi/D6zek91qfoM/hqdefault.jpg",
              width: 480,
              height: 360
            }
          },
          channelTitle: "Chris Kincade Media",
          liveBroadcastContent: "none"
        }
      }
    ]
  };

  const errorApiResponse = {
    errors: [
      {
        domain: "usageLimits",
        reason: "keyInvalid",
        message: "Bad Request"
      }
    ],
    code: 400,
    message: "Bad Request"
  };

  it("Should successfully load videos", async () => {
    videoMapUtils.getVideosAPI = jest
      .fn()
      .mockImplementation(() => Promise.resolve(videosAPIResponse));
    const action = loadVideos();
    await action(mockStore.dispatch);
    expect(mockStore.getActions()).toMatchSnapshot();
  });

  it("Should fail to load videos", async () => {
    videoMapUtils.getVideosAPI = jest
      .fn()
      .mockImplementation(() => Promise.reject(errorApiResponse));
    const action = loadVideos();
    await action(mockStore.dispatch);
    expect(mockStore.getActions()).toMatchSnapshot();
  });
});
