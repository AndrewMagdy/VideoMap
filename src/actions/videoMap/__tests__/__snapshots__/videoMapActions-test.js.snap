// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoMap action creators Should fail to load videos 1`] = `
Array [
  Object {
    "type": "LOAD_VIDEOS_REQUEST",
  },
  Object {
    "payload": Object {
      "error": "Bad Request",
    },
    "type": "LOAD_VIDEOS_ERROR",
  },
]
`;

exports[`VideoMap action creators Should successfully load videos 1`] = `
Array [
  Object {
    "type": "LOAD_VIDEOS_REQUEST",
  },
  Object {
    "payload": Object {
      "isNextPage": false,
      "videosList": Object {
        "etag": "\\"XpPGQXPnxQJhLgs6enD_n8JR4Qk/P0Q5iJsjiJ2zOzcrxDcb85guSJ8\\"",
        "items": Array [
          Object {
            "etag": "\\"XpPGQXPnxQJhLgs6enD_n8JR4Qk/ModJNYQuE9VfOulIBgXRd127_Ew\\"",
            "id": Object {
              "kind": "youtube#video",
              "videoId": "ei4r_Bs3cek",
            },
            "kind": "youtube#searchResult",
            "snippet": Object {
              "channelId": "UCVo06dBGK9VBBhq15wJxZHQ",
              "channelTitle": "Chris Kincade Media",
              "description": "Video/Edit: Chris Kincade FINE ART PRINTS www.ChrisKincade.com GET 25% OFF THE ENTIRE SITE! USE CODE “2018” AT CHECKOUT! Instagram: @Chris.",
              "liveBroadcastContent": "none",
              "publishedAt": "2018-04-26T07:43:02.000Z",
              "thumbnails": Object {
                "default": Object {
                  "height": 90,
                  "url": "https://i.ytimg.com/vi/ei4r_Bs3cek/default.jpg",
                  "width": 120,
                },
                "high": Object {
                  "height": 360,
                  "url": "https://i.ytimg.com/vi/ei4r_Bs3cek/hqdefault.jpg",
                  "width": 480,
                },
                "medium": Object {
                  "height": 180,
                  "url": "https://i.ytimg.com/vi/ei4r_Bs3cek/mqdefault.jpg",
                  "width": 320,
                },
              },
              "title": "#JEFFHUBBARD - #BODYBOARDING #PIPELINE SPECIALIST $$",
            },
          },
          Object {
            "etag": "\\"XpPGQXPnxQJhLgs6enD_n8JR4Qk/Cizv5T8I6oJtAxERYXtftJ0L3xg\\"",
            "id": Object {
              "kind": "youtube#video",
              "videoId": "D6zek91qfoM",
            },
            "kind": "youtube#searchResult",
            "snippet": Object {
              "channelId": "UCVo06dBGK9VBBhq15wJxZHQ",
              "channelTitle": "Chris Kincade Media",
              "description": "When Pipe gets too big, it's always the 662 Ride Shop boys that tackle it anyways. Here's some clips from January 22nd of Andre Botha, Trevor Kam and ...",
              "liveBroadcastContent": "none",
              "publishedAt": "2015-02-02T05:51:33.000Z",
              "thumbnails": Object {
                "default": Object {
                  "height": 90,
                  "url": "https://i.ytimg.com/vi/D6zek91qfoM/default.jpg",
                  "width": 120,
                },
                "high": Object {
                  "height": 360,
                  "url": "https://i.ytimg.com/vi/D6zek91qfoM/hqdefault.jpg",
                  "width": 480,
                },
                "medium": Object {
                  "height": 180,
                  "url": "https://i.ytimg.com/vi/D6zek91qfoM/mqdefault.jpg",
                  "width": 320,
                },
              },
              "title": "#662 Boys take XL #Pipeline - #Bodyboarding #Hawaii",
            },
          },
        ],
        "kind": "youtube#searchListResponse",
        "nextPageToken": "CAIQAA",
        "pageInfo": Object {
          "resultsPerPage": 2,
          "totalResults": 1494,
        },
        "regionCode": "EG",
      },
    },
    "type": "LOAD_VIDEOS_SUCCESS",
  },
]
`;
